### PR TITLE
Approver-led governance for Kubevirt

### LINKED ISSUE #97: Create a GOVERNANCE file
Template is [here](https://github.com/cncf/project-template/blob/main/GOVERNANCE-maintainer.md)

Community will need to approve

### PR DESCRIPTION
**Pull request by jberkus**
Creates a GOVERNANCE.md file which outlines a simple self-selecting-council governance format.

This document is based on the CNCF Maintainer governance template.

Updated 6/26, Assumptions made in the new version per @fabiand:

1. ~~"Approver" in our membership docs is equivalent to "Maintainer".~~ Maintainers become a selected position, instead of the group of Approvers to kubevirt/kubevirt
2. CoC violations will be initially reported to the approvers before being escalated to the CNCF.

Unresolved Issues:

1. We need to clean up membership_policy.md so that it reflects our actual process of contributor advancement, particularly the ability to become Approver in a subproject, and that Approver means specifically root-level approver.
2. We also need a private mailing list for the collected Maintainers, as a location to discuss security escalations.

Fixes #97

Signed-off-by: Josh Berkus <josh@agliodbs.com>

### CODE REVIEW

**Comment by fabiand** [comment]
@jberkus the file is empty.

Also adding
/hold
to get reviews from at least @davidvossel @vladikr @rmohr @mhenriks @phoracek

**Comment by jberkus** [comment]
Had a meeting with Fabian and David, text has been heavily revised based on that discussion. Please review and comment!

**Comment by EdDev** [comment]
> Assumptions made:
> 
>     1. "Approver" in our membership docs is equivalent to "Maintainer".

The template suggested here points to a single list of maintainers which are suppose to govern the whole kubevirt ecosystem.
But the original approval membership is different per how I perceive it.
The membership doc provides a guideline on how each repo should manage its contributors (except for the organization member which is not per repo). Therefore, there cannot be one maintainers list in one specific place, it is distributed across the repos.
In addition, the maintainers of individual repos are diverse, some are taking active part of many community activities and some are more focused or maintaining the correctness and quality of the code (many times with a SIG specific focus).
The description of a maintainer in this text is different, it requires much more from a maintainer.

> Unresolved Issues:
> 
>     1. We need to clean up membership_policy.md so that it reflects our actual process of contributor advancement, particularly the ability to become Approver in a subproject, and that Approver means specifically root-level approver.

Could you please clarify why it is not currently reflecting the actual process?

Regarding the root-level approvers, I think it is perfectly fine to nominate with the same "rules" SIG approvals. The reality is that in large repos (e.g. kubevirt), the code is not structured in a way that this is practical to enforce it. Therefore, an approver is added to root, but hopefully it is clear that if the nomination was given to maintain a specific SIG, the boundaries will be kept most of the times.

What is not cleared to me in the current text, is the maintainer definition. Is it over the organization? On a per-repo?

**Comment by davidvossel** [comment]
> > Assumptions made:
> > ```
> > 1. "Approver" in our membership docs is equivalent to "Maintainer".
> > ```
> 
> The template suggested here points to a single list of maintainers which are suppose to govern the whole kubevirt ecosystem.
> But the original approval membership is different per how I perceive it.
> The membership doc provides a guideline on how each repo should manage its contributors (except for the organization member which is not per repo). Therefore, there cannot be one maintainers list in one specific place, it is distributed across the repos.
> In addition, the maintainers of individual repos are diverse, some are taking active part of many community activities and some are more focused or maintaining the correctness and quality of the code (many times with a SIG specific focus).
> The description of a maintainer in this text is different, it requires much more from a maintainer.

saying an "approver" is equivalent to a KubeVirt CNCF Project Maintainer is seems inaccurate.

While it's likely that a CNCF maintainer has approver rights for the kubevirt/kubevirt repo, my understanding is that these two concepts are orthogonal to one another.

**Comment by jberkus** [comment]
@EdDev @davidvossel On 6/26, I changed the actual text to reflect a different understanding of Approver and Maintainer based on the convo with David and Fabian.  However, I just changed the PR comment now to reflect the new text.  Please re-examine, thanks.

**Comment by davidvossel** [comment]
> However, I just changed the PR comment now to reflect the new text. Please re-examine, thanks.

thanks, looks accurate now

File: GOVERNANCE.md
Comment (dankenigsberg) [inline GOVERNANCE.md:18]: Shouldn't this PR include this file, even if empty? Though I suspect we'd better initialize it with the list of people who actually do this work these days.

Instead of a new file, how about adding a `maintainers` alias into the existing OWNERS file? The function of these files would be very similar, and sure to confuse me.

File: GOVERNANCE.md
Comment (dankenigsberg) [inline GOVERNANCE.md:48]: How about, instead, doing having the proposal as a PR to the MAINTAINERS.md file?

File: GOVERNANCE.md
Comment (dankenigsberg) [inline GOVERNANCE.md:51]: can we create this list and give a URL to it here?

File: GOVERNANCE.md
Comment (dankenigsberg) [inline GOVERNANCE.md:62]: is there a quorum for these closed meetings? Does a meeting count if only one Maintainer shows up?

File: GOVERNANCE.md
Comment (dankenigsberg) [inline GOVERNANCE.md:87]:
```suggestion
* Inactivity, including 6 months or more of non-participation or non-communication,
```

File: GOVERNANCE.md
Comment (jberkus) [reply GOVERNANCE.md:62]: this doesn't supercede the requirement for a majority of maintainers to approve *anything*.  So you'd need at least 51% of the maintainers to show up.

File: GOVERNANCE.md
Comment (jberkus) [reply GOVERNANCE.md:48]: that's fine, too

File: GOVERNANCE.md
Comment (jberkus) [reply GOVERNANCE.md:51]: Apparently it already exists, added link.

File: GOVERNANCE.md
Comment (jberkus) [reply GOVERNANCE.md:18]: Dan: if you read this governance proposal, it specifically separates OWNERS for Maintainers.  That's really the meat of the proposal.  So adding the separate file instead.

**Comment by jberkus** [comment]
@dankenigsberg  That is the list we have on file with the CNCF, with the exception of Pep, whom I removed.

Supposedly Chris filed an update with them shortly before he went on medical leave, but I don't know what happened after that.

File: GOVERNANCE.md
Comment (fabiand) [inline GOVERNANCE.md:28]: Great wording!

File: GOVERNANCE.md
Comment (fabiand) [inline GOVERNANCE.md:52]: Should we have a upper limit of maintainers?

File: MAINTAINERS.md
Comment (fabiand) [inline MAINTAINERS.md:5]: Marc is awesome, but hasn't been active recently, let's drop him.

File: MAINTAINERS.md
Comment (fabiand) [inline MAINTAINERS.md:9]: Artyom is awesome, but hasn't been active recently, let's drop him.

File: MAINTAINERS.md
Comment (fabiand) [inline MAINTAINERS.md:10]: Sebastian is awesome, but hasn't been active recently, let's drop him.

File: MAINTAINERS.md
Comment (fabiand) [inline MAINTAINERS.md:11]: Marcin is awesome, but hasn't been active recently, let's drop him.

File: MAINTAINERS.md
Comment (fabiand) [inline MAINTAINERS.md:14]: I'd like to propose to add Ryan Hallisey and potentially Marcus Sorensen to the initial list of contributors.

To me both of them are already meeting the maintainer requirements.

**Review by fabiand — COMMENTED** [review]
The oevral model sounds good, specifically the separation between current kubevirt/kubevirt approvers and maintainers.

I've included a suggestion to drop some maintainers and add some others.

File: MAINTAINERS.md
Comment (fabiand) [reply MAINTAINERS.md:14]: @rthallisey @mlsorensen do you have thoughts on this?
Would you be willing to be a maintainer and take these responsibilites associated with it?

File: MAINTAINERS.md
Comment (rmohr) [inline MAINTAINERS.md:13]: I want to propose @vasiliy-ul from Suse who drove cgroupv2 for kubevirt and was quite active on bug fixing and is maintaining kubevirt for integration in ranger.

File: GOVERNANCE.md
Comment (jberkus) [reply GOVERNANCE.md:52]: You can, but I'm not sure you need one?  

The idea is that everyone who is consistently on the list of "if we're making a major project-wide change we'll have to ping this person" will be a Maintainer.  And if you have to solicit that person anyway, then it's actually easier to have them on the maintainers list.  If that ends up being a large group of people eventually, that means we have a lot of folks who are very involved in Kubevirt's future, which is hardly a bad thing.

File: MAINTAINERS.md
Comment (davidvossel) [reply MAINTAINERS.md:14]: +1 from me on these two as cncf maintainers for kubevirt

**Comment by jberkus** [comment]
@rmohr, @fabiand if you can have some kind of "official" vote/notice/whatever, I can make all those changes to the maintainer file, merge it, and tell you how to let the CNCF know.

File: MAINTAINERS.md
Comment (rmohr) [reply MAINTAINERS.md:14]: As discussed in the community meeting I would first want to see a sign of life from @mlsorensen that he will be available and responsive if needed. Apart from that, both will work from me.

File: MAINTAINERS.md
Comment (rmohr) [reply MAINTAINERS.md:13]: @davidvossel @fabiand does that work for you? Like for @mlsorensen I would want to wait for a sign of life before just blindly adding them.

**Comment by rmohr** [comment]
> @rmohr, @fabiand if you can have some kind of "official" vote/notice/whatever, I can make all those changes to the maintainer file, merge it, and tell you how to let the CNCF know.

I see that neither @mlsorensen nor @vasiliy-ul responded yet. We are in vacation time, they are probably both on vacation. I think we need a commitment for either one of those if they want to be added.

File: MAINTAINERS.md
Comment (vasiliy-ul) [reply MAINTAINERS.md:13]: Sure, I would be glad to contribute to KubeVirt in the new role if that was acceptable and fine with everyone. Probably I am still not fully familiar with all the project specifics but I am ready to dedicate additional effort to fill the gaps.

**Comment by jberkus** [comment]
I have removed the inactive maintainers per the above thread.

Per Roman and Fabian, I have added Ryan and Vasiliy as maintainers. I have not added Marcus per discussion above.

File: MAINTAINERS.md
Comment (fabiand) [reply MAINTAINERS.md:13]: Thanks for showing the openess, Vasily. We'll support you and are looking forward to your thoughts and opinions.

File: MAINTAINERS.md
Comment (fabiand) [reply MAINTAINERS.md:14]: We did not get a reply from Marcus and Ryan here, therefor I'm proposing to drop them.

**Review by fabiand — COMMENTED** [review]
Please update the reviewer list, then we should be good to go. Wohoo

**Comment by fabiand** [comment]
@jberkus I'm not seeing the maintainer changes, did you miss to push a commit?

**Comment by mazzystr** [comment]
Please add me (Chris Callegari / ***@***.*** / Red Hat) also.  This
allows me to create tickets in the CNCF Service Desk for project related
items.

**Comment by rthallisey** [comment]
Oh I didn't know I needed to reply in comment.  I gave a +1 to being added

**Review by davidvossel — COMMENTED** [review]
/lgtm

Thanks!

**Review by rmohr — APPROVED** [review]
/lgtm

**Comment by rmohr** [comment]
/approve

3 lgtm from core maintainers.

**Pull request merged**
